std/array.d

```diff
@@ -3927,3 +3927,230 @@ unittest
+
+    auto b = [0, 1].asStatic!byte;
+    assert(is(typeof(b) == byte[2]) && b == [0, 1]);
+}
+
+nothrow pure @safe unittest
+{
+    int val = 3;
+    static immutable gold = [1, 2, 3];
+    [1, 2, val].asStatic.checkStaticArray!int([1, 2, 3]);
+
+    @nogc void checkNogc()
+    {
+        [1, 2, val].asStatic.checkStaticArray!int(gold);
+    }
+
+    [1, 2, val].asStatic!double.checkStaticArray!double(gold);
+    [1, 2, 3].asStatic!int.checkStaticArray!int(gold);
+
+    [1, 2, 3].asStatic!(const(int)).checkStaticArray!(const(int))(gold);
+    [1, 2, 3].asStatic!(const(double)).checkStaticArray!(const(double))(gold);
+    {
+        const(int)[3] a2 = [1, 2, 3].asStatic;
+    }
+
+    [1, 129].asStatic!byte.checkStaticArray!byte([1, -127]);
+
+}
+
+/++
+Params: a = input range of elements
+
+Returns: A static array constructed from `a`.
++/
+auto asStatic(size_t n, T)(T a) nothrow @safe pure @nogc
+{
+    import std.conv : emplaceRef;
+    // TODO: ElementType vs ForeachType
+    alias U = typeof(a[0]);
+    U[n] ret = void;
+    size_t i;
+    foreach (ref ai; a)
+    {
+        emplaceRef!U(ret[i++], ai);
+    }
+    assert(i == n);
+    return ret;
+}
+
+/// ditto
+auto asStatic(Un : U[n], U, size_t n, T)(T a) nothrow @safe pure @nogc
+{
+    import std.conv : emplaceRef;
+    U[n] ret = void;
+    size_t i;
+    foreach (ref ai; a)
+    {
+        emplaceRef!U(ret[i++], cast(U) ai);
+    }
+    assert(i == n);
+    return ret;
+}
+
+///
+nothrow pure @safe unittest
+{
+    import std.range : iota;
+
+    auto a = 2.iota.asStatic!2;
+    assert(is(typeof(a) == int[2]) && a == [0, 1]);
+    auto b = 2.iota.asStatic!(byte[2]);
+    assert(is(typeof(b) == byte[2]) && b == [0, 1]);
+}
+
+nothrow pure @safe unittest
+{
+
+    auto a = [1, 2].asStatic;
+    assert(is(typeof(a) == int[2]) && a == [1, 2]);
+
+    import std.range : iota;
+
+    2.iota.asStatic!2.checkStaticArray!int([0, 1]);
+    2.iota.asStatic!(double[2]).checkStaticArray!double([0, 1]);
+    2.iota.asStatic!(byte[2]).checkStaticArray!byte([0, 1]);
+}
+
+nothrow pure @system unittest
+{
+    import std.range : iota;
+
+    assert(isThrown!Error(2.iota.asStatic!1));
+    assert(isThrown!Error(2.iota.asStatic!3));
+    // NOTE: alternatives are not nothrow:
+    version (none)
+    {
+        import std.exception : assertThrown;
+
+        assertThrown!Error(2.iota.asStatic!3);
+        import std.exception : ifThrown;
+
+        assert(ifThrown!Error({ 2.iota.asStatic!1; return false; }(), true));
+    }
+}
+
+@system unittest
+{
+    version (Bug)
+    {
+        // https://issues.dlang.org/show_bug.cgi?id=16779
+        auto a2 = [1, 2, 3].asStatic!byte;
+        auto a3 = [1, 2, 3].asStatic!ubyte;
+
+        // NOTE: correctly issues a deprecation
+        int[] a2 = [1, 2].asStatic;
+    }
+}
+
+/++
+Params: arr = the compile time range
+
+Returns: A static array constructed from `arr`.
++/
+auto asStatic(alias arr)() nothrow @safe pure @nogc
+{
```

Comment:
Shouldn't this entire block be in `static immutable = () {...}` to take full advantage of CTFE?

Reply:
replaced by:

```
auto asStatic(alias a)() nothrow @safe pure @nogc
{
    return .asStatic!(cast(size_t) a.length)(a);
}
```
not sure if that's what you meant but the newer form is indeed much better to avoid duplication